contrib/docker/Dockerfile

```diff
@@ -0,0 +1,320 @@
+
+	# ensure we keep apt cache around in a Docker environment
+	rm -f /etc/apt/apt.conf.d/docker-clean
+	echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
+
+	# Standard packages
+	standardPackages=(
+		apt-utils
+		ca-certificates
+		gettext-base
+		git
+		gnupg1
+		gosu
+		libcurl4-openssl-dev
+		libzip-dev
+		locales
+		locales-all
+		nano
+		procps
+		unzip
+		zip
+	)
+
+	# Image Optimization
+	imageOptimization=(
+		gifsicle
+		jpegoptim
+		optipng
+		pngquant
+	)
+
+	# Image Processing
+	imageProcessing=(
+		libjpeg62-turbo-dev
+		libmagickwand-dev
+		libpng-dev
+	)
+
+	# Required for GD
+	gdDependencies=(
+		libwebp-dev
+		libwebp6
+		libxpm-dev
+		libxpm4
+	)
+
+	# Video Processing
+	videoProcessing=(
+		ffmpeg
+	)
+
+	# Database
+	databaseDependencies=(
+		libpq-dev
+		libsqlite3-dev
+	)
+
+	apt-get update
+
+	apt-get upgrade -y
+
+	apt-get install -y --no-install-recommends \
+		${standardPackages[*]} \
+		${imageOptimization[*]} \
+		${imageProcessing[*]} \
+		${gdDependencies[*]} \
+		${videoProcessing[*]} \
+		${databaseDependencies[*]} \
+		${APT_PACKAGES_EXTRA}
+SCRIPT
+
+# update locales
+RUN set -ex \
+	&& locale-gen \
+	&& update-locale
+
+#######################################################
+# PHP: extensions
+#######################################################
+
+FROM base AS php-extensions
+
+ARG PECL_EXTENSIONS
+ARG PECL_EXTENSIONS_EXTRA
+ARG PHP_DATABASE_EXTENSIONS
+ARG PHP_DEBIAN_RELEASE
+ARG PHP_EXTENSIONS
+ARG PHP_EXTENSIONS_EXTRA
+ARG PHP_VERSION
+ARG TARGETPLATFORM
+
+RUN --mount=type=cache,id=pixelfed-php-${PHP_VERSION}-${PHP_DEBIAN_RELEASE}-${TARGETPLATFORM},sharing=locked,target=/usr/src/php/ \
+	set -ex \
+	# Grab the PHP source code so we can compile against it
+	&& docker-php-source extract \
+	# Install pecl extensions
+	&& pecl install ${PECL_EXTENSIONS} ${PECL_EXTENSIONS_EXTRA} \
+	# PHP GD extensions
+	&& docker-php-ext-configure gd \
+		--with-freetype \
+		--with-jpeg \
+		--with-webp \
+		--with-xpm \
+	# PHP extensions (dependencies)
+	&& docker-php-ext-install -j$(nproc) ${PHP_EXTENSIONS} ${PHP_EXTENSIONS_EXTRA} ${PHP_DATABASE_EXTENSIONS} \
+	# Enable all extensions
+	&& docker-php-ext-enable ${PECL_EXTENSIONS} ${PECL_EXTENSIONS_EXTRA} ${PHP_EXTENSIONS} ${PHP_EXTENSIONS_EXTRA} ${PHP_DATABASE_EXTENSIONS}
+
+#######################################################
+# PHP: composer
+#######################################################
+
+FROM base AS composer-and-src
+
+ARG PHP_VERSION
+ARG PHP_DEBIAN_RELEASE
+ARG TARGETPLATFORM
+
+# Make sure composer cache is targeting our cache mount later
+ENV COMPOSER_CACHE_DIR=/cache/composer
+
+# Don't enforce any memory limits for composer
+ENV COMPOSER_MEMORY_LIMIT=-1
+
+# Disable interactvitity from composer
+ENV COMPOSER_NO_INTERACTION=1
+
+# Copy composer from https://hub.docker.com/_/composer
+COPY --link --from=composer-image /usr/bin/composer /usr/bin/composer
+
+#! Changing user to 33
+USER 33:33
```

Comment:
TODO: document the use of numeric values rather than name (www-data) is due to how `--link` works 
TODO: possibly make these ARGS with human name for easier readability

Reply:
done!